### PR TITLE
nixosTests.networking: refactor and add NetworkManager support

### LINKED ISSUE #71151: nixos testing for NetworkManager
**Describe the bug**
NetworkManager is a particularly important component and we don't have any tests for it that could block a channel advance.  

In particular, a change can happen on nixos-unstable leaving users without functioning networking, and while rollbacks are possible it isn't an excuse to allow that to reach production systems. NixOS doesn't produce [a failsafe generation](https://github.com/cleverca22/nixos-configs/blob/5a52e0f94081361fa9d2cbfcd4f7722d442486c7/rescue_boot.nix) either.

Manual testing is also cumbersome if we want to make a new stable release.

Upstream NetworkManager does have some integration testing
* https://github.com/NetworkManager/NetworkManager-ci

but obviously not useful to us.

I think to start, if we at least had a test that verifies the services start without issues would be great.

### PR DESCRIPTION
**Pull request by Janik-Haag**
## Description of changes

This adds nixos vm tests for NetworkManager and simplifies the existing  networking tests for systemd-networkd and scripted networking. There is still a lot of todo's and this is a early-ish draft for now.

Fixes #71151

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/networking/networkmanager/default.nix
Comment (samueldr) [inline pkgs/tools/networking/networkmanager/default.nix:207]:
```suggestion
    tests = {
```

File: nixos/tests/networking/networkmanager.nix
Comment (samueldr) [inline nixos/tests/networking/networkmanager.nix:128]: Should this be removed? Implemented?

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (bjornfor) [inline nixos/doc/manual/release-notes/rl-2405.section.md:364]: Isn't this the type of config we call "settings" now? Any reason we're not using that name here?

Comment (Janik-Haag) [reply nixos/doc/manual/release-notes/rl-2405.section.md:364]: not really, would probably be correcter thing now since settings is the name we give rfc42 compliant options. I can add a `lib.mkRemovedOptionModule` error for extraConfig and rename it to settings.

Comment (Janik-Haag) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: fixed

**Comment by Janik-Haag** [comment]
I think this is now mergable. 
(I'll squash all the fixup commits once this was approved because if I do it before hand and force push reviewing gets annoying because github resets the interface on force push's)

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (jtojnar) [inline nixos/doc/manual/release-notes/rl-2405.section.md:364]: Both names the same.

Comment (Janik-Haag) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: whoops, fixed the copy past error.

File: nixos/modules/services/networking/networkmanager.nix
Comment (lf-) [inline nixos/modules/services/networking/networkmanager.nix:445]: hmmm i don't think it's quite a rename, this is a semantic change to be an attr set that should have a nicer message and probably stop evaluation saying how to fix it.

File: nixos/tests/networking/networkd-and-scripted.nix
Comment (lf-) [inline nixos/tests/networking/networkd-and-scripted.nix:161]: not fully sure where this moved, so unsure what changed here.

File: nixos/tests/networking/networkd-and-scripted.nix
Comment (lf-) [inline nixos/tests/networking/networkd-and-scripted.nix:195]: would it be possible to move these `with lib` changes into a separate pr that can be insta merged? or at least a separate commit? (can use git-revise, jujutsu, or similar to do that semi easily)

**Review by lf- — COMMENTED** [review]
some review done, thanks for writing these!

there's some of the refactoring stuff that could be split into "move stuff" and "change stuff" to make review a little bit easier.

File: nixos/tests/networking/networkd-and-scripted.nix
Comment (Janik-Haag) [reply nixos/tests/networking/networkd-and-scripted.nix:161]: In my opinion it doesn't make sense for a machine to ping it self if the other thest vm is able to communicate with the corresponding ip so I removed the lines where that was done.

File: nixos/modules/services/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:445]: Can you share a example of how to do that or point me to the correct function?

Comment (Ma27) [reply nixos/modules/services/networking/networkmanager.nix:445]: `mkRemovedOptionModule` with a good description?
My expectation would be otherwise that I can rename the option in my config and call it a day.

That said, perhaps it's good to keep `extraConfig` for one more release?

File: nixos/tests/networking/networkd-and-scripted.nix
Comment (Janik-Haag) [reply nixos/tests/networking/networkd-and-scripted.nix:195]: see #296500

File: nixos/modules/services/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:445]: > That said, perhaps it's good to keep extraConfig for one more release?

I can almost certainly guarantee you that everyone will just forgot about it until then, so I would either do it now or leave it as is, but I don't have hard opinions on this.

> mkRemovedOptionModule with a good description?

sounds good to me.

Comment (Ma27) [reply nixos/modules/services/networking/networkmanager.nix:445]: > I can almost certainly guarantee you that everyone will just forgot about it until then

I mean, it should give an annoying warning to make sure that this won't happen ;)

Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix]: I changed it to `mkRemovedOptionModule` (not sure if the markdown works there).
> > I can almost certainly guarantee you that everyone will just forgot about it until then

> I mean, it should give an annoying warning to make sure that this won't happen ;)

I meant if we do not rename extraConfig to settings that no body will remember to rename it later.
But I probably just misunderstood: "That said, perhaps it's good to keep extraConfig for one more release?"

**Review by lf- — APPROVED** [review]
thanks for writing this~

File: nixos/modules/services/networking/networkmanager.nix
Comment (Aleksanaa) [inline nixos/modules/services/networking/networkmanager.nix:174]: I don't know if this is a de facto standard, but even if many modules are moved to `settings`, often they also retain `extraConfig`. Some new modules are still adding `extraConfig`.

The benefit of this is obvious: we cannot predict that we will always generate configs for users that meet their exact needs. Some exceptions may occur inadvertently, so a guaranteed option is always good.

Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:174]: Changing to settings is the preferred key if every value can be defined using nix data structures and it's the "RFC42 style". Changing the name also doesn't really matter here since the option would break either way because the datatype was changed.

**Comment by lf-** [comment]
fixes: https://github.com/NixOS/nixpkgs/issues/71151

File: nixos/modules/services/networking/networkmanager.nix
Comment (lf-) [reply nixos/modules/services/networking/networkmanager.nix:174]: it would be possible to say, have extraConfig that gets smacked on the very end of the file, which is what `nix.settings` did, in spite of having structured config as the primary method, to allow for migration. Personally I am fine either way, the only real argument to not breaking it immediately and hard is to give time to migrate (especially for any downstream modules that need to be cross compatible across NixOS versions), but I don't think this is terrifically critical.

Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:174]: Agreed + parsing a ini and converting it to nix is fairly trivial for any usecase where the config is more then like 5 lines, I basically automated that in https://github.com/Janik-Haag/nm2nix/pull/6

File: nixos/tests/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/tests/networking/networkmanager.nix:1]: I don't think so since we I didn't write a testcase where a specific service tries to reach something on the remote. But it sounds reasonable to add something like this.

File: nixos/tests/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/tests/networking/networkmanager.nix:122]: Uhm sure we can also add a test case for checking if dns actually works. This test case was born out of #291868 which motivated me to write NM tests because I really did not want to manually test stuff every time someone opens a issue related to the NM module.

File: nixos/tests/networking/networkd-and-scripted.nix
Comment (Janik-Haag) [reply nixos/tests/networking/networkd-and-scripted.nix:1]: yeah the changes there should be identical and disappear once #296500 is merged and I can rebase this pr.

File: nixos/modules/services/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix]: fixed

File: nixos/tests/networking/networkd-and-scripted.nix
Comment (Janik-Haag) [reply nixos/tests/networking/networkd-and-scripted.nix:1]: I rebased now that the other pr is merged and the changes are gone so I'll mark this as resolved.

**Review by lilyinstarlight — APPROVED** [review]
Diff looks good to me! I see there is an open review comment about end-to-end DNS checks, but that can probably be done later in a follow-up so this PR can go ahead and get merged

I'll merge it when ofborg is green

**Comment by Janik-Haag** [comment]
> Diff looks good to me! I see there is an open review comment about end-to-end DNS checks, but that can probably be done later in a follow-up so this PR can go ahead and get merged

yes, there is two open comments about more tests:
- https://github.com/NixOS/nixpkgs/pull/292472#discussion_r1543904457
- https://github.com/NixOS/nixpkgs/pull/292472#discussion_r1543899085

but I want to add them in another pr because I want to get this in before 24.05 branch-of (because of the NM module changes)

> I'll merge it when ofborg is green

thank you <3
you can also build `nix-build -A nixosTests.networking.*` if you want to be 100% sure (afaik ofborg only builds vm tests if a package with passthru.tests get's modified)

File: nixos/tests/networking/networkmanager.nix
Comment (lf-) [inline nixos/tests/networking/networkmanager.nix:80]: oh gods. i am sorry i didn't notice this earlier. this is broken when `network-online.target` is not depended on by multi-user.target.
```suggestion
        router.systemctl("start network-online.target")
        router.wait_for_unit("network-online.target")
```

File: nixos/tests/networking/networkmanager.nix
Comment (lf-) [inline nixos/tests/networking/networkmanager.nix:53]:
```suggestion
        router.systemctl("start network-online.target")
        router.wait_for_unit("network-online.target")
```

File: nixos/tests/networking/networkmanager.nix
Comment (lf-) [inline nixos/tests/networking/networkmanager.nix:116]:
```suggestion
        router.systemctl("start network-online.target")
        router.wait_for_unit("network-online.target")
```

**Review by lf- — CHANGES_REQUESTED** [review]

File: nixos/tests/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/tests/networking/networkmanager.nix:82]: I think it's fixed now. Thank's for the headsup!

**Comment by lf-** [comment]
Builds on my machine, and the changes are trivial enough to trust the last ofborg run.  Merging.

**Review by lf- — APPROVED** [review]

**Pull request merged**
